### PR TITLE
ml services version 2.0.0

### PR DESCRIPTION
**Pull request by GitHK**
# Changes

- ported `jupyter-ml-tensorflow` and `jupyter-ml-pytorch` to use `dynamic-sidecar`
- upgraded tensorflow to version 2.9.0
- upgraded pytorch to version 1.11.0
- enabled LaTex compiling
- enabled voila mode and voila preview
- upgraded python to `3.9.12`

### CODE REVIEW

**Review by elisabettai — APPROVED** [review]
Looks great to me!

File: .osparc/jupyter-ml-pytorch/metadata.yml [hunk 1]
```diff
@@ -0,0 +1,71 @@
+name: Jupyter ML PyTorch
+thumbnail: https://imgur.com/0ndE7uX.png
+description: "JupyterLab with PyTorch installed. No other scientific tools are provided."
+key: simcore/services/dynamic/jupyter-ml-pytorch
+version: 2.0.0
+integration-version: "2.0.0"
+type: dynamic
+authors:
+  - name: "Andrei Neagu"
+    email: "neagu@itis.swiss"
```
Comment (mguidon) [inline .osparc/jupyter-ml-pytorch/metadata.yml:10]: quotes vs no quotes for fasse

File: .osparc/jupyter-ml-pytorch/runtime.yml [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    value:
+      Limits:
+        NanoCPUs: 4000000000
+        MemoryBytes: 17179869184
+      Reservations:
+        NanoCPUs: 100000000
+        MemoryBytes: 536870912
+        GenericResources:
+          - DiscreteResourceSpec:
+              Kind: AIRAM
```
Comment (mguidon) [inline .osparc/jupyter-ml-pytorch/runtime.yml:31]: Once they are in prod I would like to see how the behave. We might be able to get rid of AIRAM

File: .osparc/jupyter-ml-pytorch/runtime.yml [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    value: 8888
+  - name: env
+    type: string
+    value:
+      - DISPLAY=:0
+  - name: mount
+    type: object
+    value:
+      - ReadOnly: true
+        Source: /tmp/.X11-unix
```
[the remark this answers is not in the document]
Comment (GitHK) [reply .osparc/jupyter-ml-pytorch/runtime.yml:13]: Yes, it was already there in the older version

File: .osparc/jupyter-ml-pytorch/runtime.yml [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+        NanoCPUs: 4000000000
+        MemoryBytes: 17179869184
+      Reservations:
+        NanoCPUs: 100000000
+        MemoryBytes: 536870912
+        GenericResources:
+          - DiscreteResourceSpec:
+              Kind: AIRAM
+              Value: 1
+restart-policy: no-restart
```
Comment (mguidon) [inline .osparc/jupyter-ml-pytorch/runtime.yml:33]: no-restart?

File: Makefile [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+	@make compose-spec
+	@$(call _bumpversion,$<,version-pytorch-)
+	@make compose-spec
+
+.PHONY: compose-spec
+compose-spec: ## runs ooil to assemble the docker-compose.yml file
+	@docker run -it --rm -v $(PWD):/iseg \
+		-u $(shell id -u):$(shell id -g) \
+		itisfoundation/ci-service-integration-library:v1.0.1-dev-31 \
+		sh -c "cd /iseg && ooil compose"
```
Comment (mguidon) [inline Makefile:34]: iseg?

File: .osparc/jupyter-ml-pytorch/runtime.yml (hunk 4 above)
Comment (GitHK) [reply .osparc/jupyter-ml-pytorch/runtime.yml:33]: Something optional (it has to do with UI services which require a reboot after the inputs are pulled), by default is set to `no-restart`.

File: Makefile [hunk 6]
```diff
@@ -0,0 +1,63 @@
+SHELL = /bin/sh
+.DEFAULT_GOAL := help
+
+export IMAGE_PYTORCH=jupyter-ml-pytorch
+export IMAGE_TENSORFLOW=jupyter-ml-tensorflow
+export TAG_PYTORCH=2.0.0
```
Comment (mguidon) [inline Makefile:6]: Can the bumpversion container handle two TAGs with the same version?

File: .osparc/jupyter-ml-pytorch/metadata.yml (hunk 1 above)
Comment (GitHK) [reply .osparc/jupyter-ml-pytorch/metadata.yml:10]: Sorry he is already here. I would keep him.

File: Makefile (hunk 6 above)
Comment (GitHK) [reply Makefile:6]: yes, the trick was adding `"` to the integration version

File: common/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,293 @@
+FROM nvidia/cuda:11.0.3-cudnn8-runtime-ubuntu18.04 as ml-base
```
[the remark this answers is not in the document]
Comment (GitHK) [reply common/Dockerfile:1]: Yes, for now I've ported more or less the same as previously defined.

File: common/docker/boot_notebook.bash [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+        MAX_CPUS=1
+    fi
+    echo "$INFO" "Setting Z43_MAX_CPU_RESOURCES to " "${MAX_CPUS}"
+    export Z43_MAX_CPU_RESOURCES="${MAX_CPUS}"
+    export OMP_NUM_THREADS="${MAX_CPUS}"
+fi
+
+## set licensing information if not yet set via platform
+PORT=27000
+declare -a SERVERS=("license.speag.com" "license-1.speag.com" "10.1.0.21")
```
Comment (mguidon) [inline common/docker/boot_notebook.bash:75]: same here. Licensing only makes sense for sim4life

File: common/docker/init.bash [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+# http://redsymbol.net/articles/unofficial-bash-strict-mode/
+set -euo pipefail
+IFS=$'\n\t'
+
+INFO="INFO: [init.bash] "
+
+echo "$INFO" "Starting jupyter-smash init.bash"
+
+
+if [ -x "$(command -v nvidia-smi)" ]; then
```
Comment (mguidon) [inline common/docker/init.bash:11]: Again, this is for iSolve only

File: common/Dockerfile [hunk 10]
```diff
@@ -0,0 +1,293 @@
...
+
+# Install conda as jovyan and check the md5 sum provided on the download site
+ENV MINICONDA_VERSION=4.8.2 \
+    MINICONDA_MD5=87e77f097f6ebb5127c77662dfc3165e \
+    CONDA_VERSION=4.8.2
+
+RUN cd /tmp && \
+    wget --quiet https://repo.continuum.io/miniconda/Miniconda3-py37_${MINICONDA_VERSION}-Linux-x86_64.sh && \
+    echo "${MINICONDA_MD5} *Miniconda3-py37_${MINICONDA_VERSION}-Linux-x86_64.sh" | md5sum -c - && \
+    /bin/bash Miniconda3-py37_${MINICONDA_VERSION}-Linux-x86_64.sh -f -b -p $CONDA_DIR && \
```
[the remark this answers is not in the document]
Comment (GitHK) [reply common/Dockerfile:79]: That's why I ended up creating a new ipythonkernel

File: common/docker/boot_notebook.bash [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+        "webbrowser_open_new": 0,
+        "notebook_dir": "${NOTEBOOK_BASE_DIR}",
+        "root_dir": "${NOTEBOOK_BASE_DIR}",
+        "preferred_dir": "${NOTEBOOK_BASE_DIR}/workspace/"
+    }
+}
+EOF
+
+# Checking for resources limit env vars injected by osparc
+SIMCORE_NANO_CPUS_LIMIT="${SIMCORE_NANO_CPUS_LIMIT:-0}"
```
[the remark this answers is not in the document]
Comment (GitHK) [reply common/docker/boot_notebook.bash:57]: Will remove this, if not required

File: common/docker/boot_notebook.bash (hunk 8 above)
Comment (GitHK) [reply common/docker/boot_notebook.bash:75]: will remove also this

File: docker-compose-local.yml [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+    image: simcore/services/dynamic/${IMAGE_TO_RUN}:${TAG_TO_RUN}
+    ports:
+      - 8888:8888
+    environment:
+      - DY_SIDECAR_PATH_INPUTS=/tmp/inputs
+      - DY_SIDECAR_PATH_OUTPUTS=/tmp/outputs
+      - DY_BOOT_OPTION_BOOT_MODE=0
+      - DY_SIDECAR_PATH/home/jovyan/work/workspace
+    volumes:
+      - /tmp/.X11-unix:/tmp/.X11-unix
```
Comment (mguidon) [inline docker-compose-local.yml:13]: Again, I doubt we need an X server (smash does though)

Comment (GitHK) [reply docker-compose-local.yml:13]: let me try and drop it, see what happens

**Review by mguidon — CHANGES_REQUESTED** [review]
Please remove the copy/paste parts from jupyter-smash. Otherwise looking good.

File: .osparc/jupyter-ml-pytorch/metadata.yml (hunk 1 above)
Comment (mguidon) [reply .osparc/jupyter-ml-pytorch/metadata.yml:10]: I meant "Andrei Neagu" vs Alessandro Fasse. (The quotes)

**Review by mguidon — APPROVED** [review]

**Pull request merged**